File: src/orm/data-types.ts
export enum DataType {
    STRING = 'VARCHAR',
    CHAR = 'CHAR',
    INTEGER = 'INTEGER',
    SMALLINT = 'SMALLINT',
    BIGINT = 'BIGINT',
    NUMERIC = 'NUMERIC',
    FLOAT = 'FLOAT',
    DOUBLE = 'DOUBLE PRECISION',
    TEXT = 'BLOB_TEXT',   // Para campos de texto longo (Sub_type 1)
    BINARY = 'BLOB_BIN',  // Para assinaturas digitais/imagens (Sub_type 0)
    BOOLEAN = 'BOOLEAN',
    DATE = 'DATE',
    TIME = 'TIME',
    DATEONLY = 'DATE',
    TIMESTAMP = 'TIMESTAMP',
    DECIMAL = 'DECIMAL',
    ENUM = 'ENUM'
}

export type ValidatorFn = (value: any) => boolean | string | Promise<boolean | string>;

export interface DataTypeDefinition {
    key: string;
    sql: string;
    length?: number;
    precision?: number;
    scale?: number;
    values?: Array<string | number>;
}

export type DataTypeInput = DataType | DataTypeDefinition;

export interface ModelReferenceLike {
    modelName?: string;
    tableName?: string;
    name?: string;
}

export interface ColumnReferenceOptions {
    model: string | ModelReferenceLike | Function;
    key: string;
    /** Nome do constraint FK no banco; se ausente, é gerado em `sync`. */
    constraintName?: string;
}

export interface ColumnOptions {
    type: DataTypeInput;
    primaryKey?: boolean;
    autoIncrement?: boolean;
    sequence?: string; // Nome do Generator/Sequence no Firebird
    allowNull?: boolean;
    defaultValue?: any;
    unique?: boolean;
    references?: ColumnReferenceOptions;
    onUpdate?: 'CASCADE' | 'RESTRICT' | 'NO ACTION' | 'SET NULL' | 'SET DEFAULT';
    onDelete?: 'CASCADE' | 'RESTRICT' | 'NO ACTION' | 'SET NULL' | 'SET DEFAULT';
    /**
     * Função única ou registo: funções `ValidatorFn` e/ou **validadores built-in** (ex.: `isEmail: true`,
     * `len: [0, 255]`, `isIn: [['A','B']]`, `not: { isEmail: true }` — ver `BUILTIN_VALIDATOR_NAMES` em
     * `validators.ts`).
     */
    validate?: Record<string, ValidatorFn | unknown> | ValidatorFn;
}

function createDefinition(definition: DataTypeDefinition): DataTypeDefinition {
    return Object.freeze(definition);
}

export const DataTypes = {
    STRING(length = 255): DataTypeDefinition {
        return createDefinition({ key: 'STRING', sql: `VARCHAR(${length})`, length });
    },
    CHAR(length = 1): DataTypeDefinition {
        return createDefinition({ key: 'CHAR', sql: `CHAR(${length})`, length });
    },
    INTEGER(): DataTypeDefinition {
        return createDefinition({ key: 'INTEGER', sql: 'INTEGER' });
    },
    SMALLINT(): DataTypeDefinition {
        return createDefinition({ key: 'SMALLINT', sql: 'SMALLINT' });
    },
    BIGINT(): DataTypeDefinition {
        return createDefinition({ key: 'BIGINT', sql: 'BIGINT' });
    },
    NUMERIC(precision = 18, scale = 0): DataTypeDefinition {
        return createDefinition({ key: 'NUMERIC', sql: `NUMERIC(${precision}, ${scale})`, precision, scale });
    },
    FLOAT(): DataTypeDefinition {
        return createDefinition({ key: 'FLOAT', sql: 'FLOAT' });
    },
    DOUBLE(): DataTypeDefinition {
        return createDefinition({ key: 'DOUBLE', sql: 'DOUBLE PRECISION' });
    },
    TEXT(): DataTypeDefinition {
        return createDefinition({ key: 'TEXT', sql: 'BLOB SUB_TYPE TEXT' });
    },
    BLOB(): DataTypeDefinition {
        return createDefinition({ key: 'BLOB', sql: 'BLOB' });
    },
    BOOLEAN(): DataTypeDefinition {
        return createDefinition({ key: 'BOOLEAN', sql: 'BOOLEAN' });
    },
    DATE(): DataTypeDefinition {
        return createDefinition({ key: 'DATE', sql: 'DATE' });
    },
    TIME(): DataTypeDefinition {
        return createDefinition({ key: 'TIME', sql: 'TIME' });
    },
    DATEONLY(): DataTypeDefinition {
        return createDefinition({ key: 'DATEONLY', sql: 'DATE' });
    },
    TIMESTAMP(): DataTypeDefinition {
        return createDefinition({ key: 'TIMESTAMP', sql: 'TIMESTAMP' });
    },
    DECIMAL(precision = 18, scale = 2): DataTypeDefinition {
        return createDefinition({ key: 'DECIMAL', sql: `DECIMAL(${precision}, ${scale})`, precision, scale });
    },
    ENUM(...values: Array<string | number>): DataTypeDefinition {
        return createDefinition({ key: 'ENUM', sql: 'VARCHAR(255)', values });
    }
};

export function isTextType(type: DataTypeInput): boolean {
    if (typeof type === 'string') return type === DataType.TEXT;
    return type.key.toUpperCase() === 'TEXT';
}

File: src/orm_py/data_types.py
from __future__ import annotations

from enum import Enum
from typing import Any, Callable, Dict, List, Optional, TypedDict, Union
from sqlalchemy.types import UserDefinedType


class DataType(str, Enum):
    STRING = "VARCHAR"
    CHAR = "CHAR"
    INTEGER = "INTEGER"
    SMALLINT = "SMALLINT"
    BIGINT = "BIGINT"
    NUMERIC = "NUMERIC"
    FLOAT = "FLOAT"
    DOUBLE = "DOUBLE PRECISION"
    TEXT = "BLOB_TEXT"
    BINARY = "BLOB_BIN"
    BOOLEAN = "BOOLEAN"
    DATE = "DATE"
    TIME = "TIME"
    DATEONLY = "DATE"
    TIMESTAMP = "TIMESTAMP"
    DECIMAL = "DECIMAL"
    ENUM = "ENUM"
    BLOB = "BLOB"
    BLOB_BINARY = "BLOB SUB_TYPE BINARY"


class BlobSubTypeText(UserDefinedType):
    def get_col_spec(self, **kw: Any) -> str:
        return "BLOB SUB_TYPE TEXT"


class BlobSubTypeBinary(UserDefinedType):
    def get_col_spec(self, **kw: Any) -> str:
        return "BLOB SUB_TYPE BINARY"


class DataTypeDefinition(TypedDict, total=False):
    key: str
    sql: str
    length: int
    precision: int
    scale: int
    values: List[Union[str, int]]


DataTypeInput = Union[DataType, DataTypeDefinition, str]
ValidatorFn = Callable[[Any], Union[bool, str]]


class ColumnReferenceOptions(TypedDict, total=False):
    model: Union[str, object]
    key: str
    constraintName: str


class ColumnOptions(TypedDict, total=False):
    type: DataTypeInput
    primaryKey: bool
    autoIncrement: bool
    sequence: str
    allowNull: bool
    defaultValue: Any
    unique: bool
    references: ColumnReferenceOptions
    onUpdate: str
    onDelete: str
    validate: Union[Dict[str, Union[ValidatorFn, Any]], ValidatorFn]


class DataTypes:
    @staticmethod
    def STRING(length: int = 255) -> DataTypeDefinition:
        return {"key": "STRING", "sql": f"VARCHAR({length})", "length": length}

    @staticmethod
    def CHAR(length: int = 1) -> DataTypeDefinition:
        return {"key": "CHAR", "sql": f"CHAR({length})", "length": length}

    @staticmethod
    def INTEGER() -> DataTypeDefinition:
        return {"key": "INTEGER", "sql": "INTEGER"}

    @staticmethod
    def SMALLINT() -> DataTypeDefinition:
        return {"key": "SMALLINT", "sql": "SMALLINT"}

    @staticmethod
    def BIGINT() -> DataTypeDefinition:
        return {"key": "BIGINT", "sql": "BIGINT"}

    @staticmethod
    def NUMERIC(precision: int = 18, scale: int = 0) -> DataTypeDefinition:
        return {"key": "NUMERIC", "sql": f"NUMERIC({precision}, {scale})", "precision": precision, "scale": scale}

    @staticmethod
    def FLOAT() -> DataTypeDefinition:
        return {"key": "FLOAT", "sql": "FLOAT"}

    @staticmethod
    def DOUBLE() -> DataTypeDefinition:
        return {"key": "DOUBLE", "sql": "DOUBLE PRECISION"}

    @staticmethod
    def TEXT() -> DataTypeDefinition:
        return {"key": "TEXT", "sql": "BLOB SUB_TYPE TEXT"}

    @staticmethod
    def BLOB() -> DataTypeDefinition:
        return {"key": "BLOB", "sql": "BLOB"}

    @staticmethod
    def BLOB_TEXT() -> DataTypeDefinition:
        return {"key": "BLOB_TEXT", "sql": "BLOB SUB_TYPE TEXT"}

    @staticmethod
    def BLOB_BINARY() -> DataTypeDefinition:
        return {"key": "BLOB_BINARY", "sql": "BLOB SUB_TYPE BINARY"}

    @staticmethod
    def BOOLEAN() -> DataTypeDefinition:
        return {"key": "BOOLEAN", "sql": "BOOLEAN"}

    @staticmethod
    def DATE() -> DataTypeDefinition:
        return {"key": "DATE", "sql": "DATE"}

    @staticmethod
    def TIME() -> DataTypeDefinition:
        return {"key": "TIME", "sql": "TIME"}

    @staticmethod
    def DATEONLY() -> DataTypeDefinition:
        return {"key": "DATEONLY", "sql": "DATE"}

    @staticmethod
    def TIMESTAMP() -> DataTypeDefinition:
        return {"key": "TIMESTAMP", "sql": "TIMESTAMP"}

    @staticmethod
    def DECIMAL(precision: int = 18, scale: int = 2) -> DataTypeDefinition:
        return {"key": "DECIMAL", "sql": f"DECIMAL({precision}, {scale})", "precision": precision, "scale": scale}

    @staticmethod
    def ENUM(*values: Union[str, int]) -> DataTypeDefinition:
        return {"key": "ENUM", "sql": "VARCHAR(255)", "values": list(values)}


def is_text_type(tp: DataTypeInput) -> bool:
    if isinstance(tp, dict):
        return str(tp.get("key", "")).upper() == "TEXT"
    return str(tp).upper() in {"TEXT", "BLOB_TEXT"}

